Transformers blog post (#95)
rstudio/ai-blog: _posts/2020-07-28-state-of-the-art-nlp-models-from-r/state-of-the-art-nlp-models-from-r.Rmd

    background-repeat: no-repeat;
    background-size: 51px;
    background-position-y: -4px;
    display: inline-block;
    padding-left: 24px;
    border-radius: 4px;
    text-decoration: none;
}
</style>

```{r setup, include=FALSE, eval=F,echo=T}
knitr::opts_chunk$set(echo = FALSE, eval = F, echo = T)
```

## Introduction

The _Transformers_ repository from ["Hugging Face"](https://github.com/huggingface/transformers) contains a lot of ready to use, state-of-the-art models, which are straightforward to download and fine-tune with Tensorflow & Keras. 

For this purpose the users usually need to get:

* The model itself (e.g. Bert, Albert, RoBerta, GPT-2 and etc.)
* The tokenizer object
* The weights of the model

In this post, we will work on a classic binary classification task and train our dataset on 3 models: 

* [GPT-2](https://blog.openai.com/better-language-models/) from Open AI
* [RoBERTa](https://arxiv.org/abs/1907.11692) from Facebook
* [Electra](https://arxiv.org/abs/2003.10555) from Google Research/Stanford University

However, readers should know that one can work with transformers on a variety of down-stream tasks, such as:

> 1) feature-extraction

Reviewer: ```suggestion
1) feature extraction
```

> 2) sentiment-analysis

Reviewer: ```suggestion
2) sentiment analysis
```

> 3) [text-classification](https://github.com/huggingface/transformers/tree/master/examples/text-classification)

Reviewer: ```suggestion
3) [text classification](https://github.com/huggingface/transformers/tree/master/examples/text-classification)
```

> 4) [question-answering](https://github.com/huggingface/transformers/tree/master/examples/question-answering)

Reviewer: ```suggestion
4) [question answering](https://github.com/huggingface/transformers/tree/master/examples/question-answering)
```

  do.call(rbind,gather_history[[x]][["metrics"]]) %>% 
    as.data.frame() %>% 
    tibble::rownames_to_column() %>% 
    mutate(model_names = names(gather_history[x])) 
}, simplify = F) %>% do.call(plyr::rbind.fill,.) %>% 
  mutate(rowname = stringr::str_extract(rowname, 'loss|val_loss|auc|val_auc')) %>% 
  rename(epoch_1 = V1, epoch_2 = V2)
```

```{r eval=T,echo=F}
library(dplyr)
res = data.table::fread('files/res.csv') %>% 
  filter(rowname %in% 'val_auc') %>% arrange(desc(V2)) %>% 
   rename(epoch_1 = V1, epoch_2 = V2, metric = rowname) %>% 
  mutate(epoch_1 = round(epoch_1,3),epoch_2 = round(epoch_2,3))
DT::datatable(res, options = list(dom = 't'))
```

Both the _RoBERTa_ and _Electra_ models show some additional improvements after 2 epochs of training, which cannot be said of  _GPT-2_. In this case, it is clear that it can be enough to train a state-of-the-art model even for a single epoch.

## Conclusion

In this post, we showed how to use state-of-the-art NLP models from R. 
To understand how to apply them to more complex tasks, it is highly recommended to review the [transformers tutorial](https://github.com/huggingface/transformers/tree/master/examples).

> We encourage readers to try out these models and share the benchmarks below in the comments section!

Reviewer: ```suggestion
We encourage readers to try out these models and share their results below in the comments section!
```

---
title: "State-of-the-art NLP models from R"
description: |

>   Nowadays, Microsoft, Google, Facebook, and OpenAI share lots of state-of-the-art models in the field of Natural Language Processing. However, there are fewer materials for the practical application of these models for the R community. In this post, we will show how the R users can access and benefit these models as well.

Reviewer: ```suggestion
  Nowadays, Microsoft, Google, Facebook, and OpenAI are sharing lots of state-of-the-art models in the field of Natural Language Processing. However, fewer materials exist how to use these models from R. In this post, we will show how R users can access and benefit from these models as well.
```

    affiliation: QSS Analytics
    affiliation_url: http://www.qss.az/
date: 07-28-2020
categories:
  - Natural Language Processing
creative_commons: CC BY
repository_url: https://github.com/henry090/transformers
output: 
  distill::distill_article:
    self_contained: false
    toc: true
    toc_depth: 2
preview: files/dino.gif
---

<style type="text/css">
.colab-root {
    display: inline-block;
    background: rgba(255, 255, 255, 0.75);
    padding: 4px 8px;
    border-radius: 4px;
    font-size: 11px!important;
    text-decoration: none;
    color: #aaa;
    border: none;
    font-weight: 300;
    border: solid 1px rgba(0, 0, 0, 0.08);
    border-bottom-color: rgba(0, 0, 0, 0.15);
    text-transform: uppercase;
    line-height: 16px;
}
span.colab-span {
    background-image: url(https://www.vectorlogo.zone/logos/kaggle/kaggle-ar21.svg);
    background-repeat: no-repeat;
    background-size: 51px;
    background-position-y: -4px;
    display: inline-block;
    padding-left: 24px;
    border-radius: 4px;
    text-decoration: none;
}
</style>

```{r setup, include=FALSE, eval=F,echo=T}
knitr::opts_chunk$set(echo = FALSE, eval = F, echo = T)
```

## Introduction

> Transformers repository from ["Hugging Face"](https://github.com/huggingface/transformers) contains a lot of ready to use state-of-the-art models which is very straightforward to download and fine-tune with Tensorflow & Keras. 

Reviewer: ```suggestion
The _Transformers_ repository from ["Hugging Face"](https://github.com/huggingface/transformers) contains a lot of ready to use, state-of-the-art models, which are straightforward to download and fine-tune with Tensorflow & Keras. 
```

For this purpose the users usually need to get:

* The model itself (e.g. Bert, Albert, RoBerta, GPT-2 and etc.)
* The tokenizer object
* The weights of the model

> In this post, we will work on classic binary classification task and train our data set on 3 models: 

Reviewer: ```suggestion
In this post, we will work on a classic binary classification task and train our dataset on 3 models: 
```

* [GPT-2](https://blog.openai.com/better-language-models/) from Open AI
* [RoBERTa](https://arxiv.org/abs/1907.11692) from Facebook
* [Electra](https://arxiv.org/abs/2003.10555) from Google Research/Stanford University

> However, readers should know that one can work with transformers on a variety type down-stream tasks, such as:

Reviewer: ```suggestion
However, readers should know that one can work with transformers on a variety of down-stream tasks, such as:
```

Reviewer: I suggest adding a footnote with a few links to tutorials that show how to use these models for downstream tasks.

Author: I added hyperlinks

1) feature-extraction
2) sentiment-analysis
3) text-classification
4) question-answering
5) summarization
6) translation and many more.

## Prerequisites

> Our first job is to install _transformers_ package via ```reticulate```.

Reviewer: ```suggestion
Our first job is to install the _transformers_ package via ```reticulate```.
```

```{r eval = F, echo = T}
reticulate::py_install('transformers', pip = TRUE)
```

> Then as usual call standard 'Keras', 'TensorFlow' >= 2.0 and some classic libraries from R.

Reviewer: ```suggestion
Then, as usual, load standard 'Keras', 'TensorFlow' >= 2.0 and some classic libraries from R.
```

```{r eval = F, echo = T}
library(keras)
library(tensorflow)
library(dplyr)
library(tfdatasets)

transformer = reticulate::import('transformers')
```

> Note, that for _GPU TensorFlow_ one could specify the following parameters in order to avoid any memory issues.

Reviewer: ```suggestion
Note that if running TensorFlow on GPU one could specify the following parameters in order to avoid memory issues.
```

```{r eval = F, echo = T}

physical_devices = tf$config$list_physical_devices('GPU')
tf$config$experimental$set_memory_growth(physical_devices[[1]],TRUE)

tf$keras$backend$set_floatx('float32')

```

## Template

We already mentioned that to train a data on the specific model, users should download the model, its tokenizer object and weights. For example, to get a RoBERTa model one has to do the following:

```{r eval = F, echo = T}
# get Tokenizer
transformer$RobertaTokenizer$from_pretrained('roberta-base', do_lower_case=TRUE)

# get Model with weights
transformer$TFRobertaModel$from_pretrained('roberta-base')
```

## Data preparation

> There is a ready data for binary classification in [text2vec](http://text2vec.org/) package. Let's upload the data set and take a sample for fast model training.

Reviewer: ```suggestion
A dataset for binary classification is provided in [text2vec](http://text2vec.org/) package. Let's load the dataset and take a sample for fast model training.
```

```{r eval = F, echo = T}
library(text2vec)
data("movie_review")
df = movie_review %>% rename(target = sentiment, comment_text = review) %>% 
  sample_n(2000) %>% 
  data.table::as.data.table()
```

Split our data into 2 parts:

```{r eval = F, echo = T}
idx_train = sample.int(nrow(df)*0.8)

train = df[idx_train,]
test = df[!idx_train,]
```

## Data input for Keras

> Till now, we just covered data importing and splitting part. To feed input to network we have to turn our raw text into indices via imported tokenizer. And then adapt the model to binary classification model by adding a dense layer with 1 unit at the end.

Reviewer: ```suggestion
Until now, we've just covered data import and train-test split. To feed input to the network we have to turn our raw text into indices via the imported tokenizer. And then adapt the model to do binary classification by adding a dense layer with a single unit at the end.
```

  history = model %>% keras::fit(tf_train, epochs=epochs, #steps_per_epoch=len/batch_size,
                validation_data=tf_test)
  gather_history[[i]]<- history
  names(gather_history)[i] = ai_m[[i]][1]
}

```

<center>
<a href="https://www.kaggle.com/henry090/transformers" class="colab-root">Reproduce in a<span class="colab-span">  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Notebook</span></a>
</center>
<br>

Extract results to see the benchmarks:

```{r eval = F, echo = T}
res = sapply(1:3, function(x) {
  do.call(rbind,gather_history[[x]][["metrics"]]) %>% 
    as.data.frame() %>% 
    tibble::rownames_to_column() %>% 
    mutate(model_names = names(gather_history[x])) 
}, simplify = F) %>% do.call(plyr::rbind.fill,.) %>% 
  mutate(rowname = stringr::str_extract(rowname, 'loss|val_loss|auc|val_auc')) %>% 
  rename(epoch_1 = V1, epoch_2 = V2)
```

```{r eval=T,echo=F}
library(dplyr)
res = data.table::fread('files/res.csv') %>% 
  filter(rowname %in% 'val_auc') %>% arrange(desc(V2)) %>% 
   rename(epoch_1 = V1, epoch_2 = V2, metric = rowname) %>% 
  mutate(epoch_1 = round(epoch_1,2),epoch_2 = round(epoch_2,2))
DT::datatable(res, options = list(dom = 't'))
```

> RoBERTa and Electra models have some improvements after 2 epochs what cannot be said for GPT-2 model. In this case, it is clear that it can be enough to train state-of-the-art model even for 1 epoch.

Reviewer: ```suggestion
Both the _RoBERTa_ and _Electra_ models show some additional improvements after 2 epochs of training, which cannot be said of  _GPT-2_. In this case, it is clear that it can be enough to train a state-of-the-art model even for a single epoch.
```

## Conclusion

> In this post, we showed the possibility of application of the state-of-the-art NLP models from R. 

Reviewer: ```suggestion
In this post, we showed how to use state-of-the-art NLP models from R. 
```

> To understand more complex tasks, it is highly recommended to review [transformers tutorial](https://github.com/huggingface/transformers/tree/master/examples) and see vignettes of [Keras GitHub](https://github.com/rstudio/keras/tree/master/vignettes/examples) page.

Reviewer: ```suggestion
To understand how to apply them to more complex tasks, it is highly recommended to review the [transformers tutorial](https://github.com/huggingface/transformers/tree/master/examples).
```

Reviewer: I removed the link to generic Keras vignettes, as I think it's too - well, generic ;-)
You could add some specific links instead if you have some.
Also, I'd conclude with some motivating sentence encouraging readers to try out these models, and possibly report in the comments if they had good results!

Author: I added an additional sentence here
